### DISCUSSION TITLE
How is "Stabilizer.thrust" calculated?

### DISCUSSION THREAD

**Discussion by WoodyloveBo** [bitcraze/discussions#1306] [Q&A] [answered]
Hello. I attempted to log the thrust based on motor rotation using the code provided in cflib:

[https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/motors/ramp.py](url)

However, upon running the code, I noticed that the entire thrust was logged, despite confirming that the rotation between the four motors was unbalanced.

https://github.com/bitcraze/discussions/assets/143582301/9bb2bc02-ceeb-4d05-a7a1-93f5185e5dab

I'm interested in knowing the thrust (possibly PWM) of each motor when they rotate in the same way. I believe this thrust corresponds to the input from "Stabilizer.thrust." I'm curious about how this "Stabilizer.thrust" value is calculated from the PWM of each motor.

Thank you.

**Comment by ataffanel** [comment]
The variable `stabilizer.thrust` is the thrust to be applied globally to the platform expressed in a unit-less 0-65636 (0-100%): https://github.com/bitcraze/crazyflie-firmware/blob/d9ee34a78cecf2eaf8b05e0b54487cff0c7b4590/src/modules/src/stabilizer.c#L520-L523.

It is then passed, together with roll/pitch/yaw control to the power distribution to generate actual PWM for each motors. The `motor` log group allows you to get the control on the motors: https://github.com/bitcraze/crazyflie-firmware/blob/d9ee34a78cecf2eaf8b05e0b54487cff0c7b4590/src/modules/src/stabilizer.c#L818-L843

**Reply by WoodyloveBo** [reply]
Thank you for your response ! 

As you mentioned, I understand that the Stabilizer.thrust variable is represented as a dimensionless value from 0 to 65636. What I'm curious about is where this Stabilizer.thrust variable is calculated from.

I've looked through several source files but couldn't find it.

**Reply by gemenerik** [reply]
Arnaud's first link points to where the stabilizer.thrust log param is created. There you can see it is actually linked to control.thrust. Where this is set depends on the used controller. For the PID controller (default) [`thrust` is set in `controller_pid.c`](https://github.com/bitcraze/crazyflie-firmware/blob/2023.11/src/modules/src/controller/controller_pid.c#L143). 

If you are unaware, an IDE like vscode can help you find references:
![image](https://github.com/bitcraze/discussions/assets/49898887/c4f7e30c-ba15-4a69-be89-c1936caac17d)

**Reply by WoodyloveBo** [reply]
Thank you, now I understand how the control.thrust is connected. Currently, I'm examining the distribution of thrust to the motors, and I've confirmed that there are three control modes: Legacy, ForceTorque, and Force. I would like to know which distribution method I am currently using. Where can I find this information? Or, can this be determined using Python scripts?

**Reply by ataffanel** [reply]
Can you please link to the piece of code you want explanation about? Where are you in the power distribution. If something can be set/seen by the client or python it is usually be made available as a parameter or log. Those are normally declared at the bottom of files.

**Reply by WoodyloveBo** [reply]
https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller/controller_pid.c

The **control->thrust** mentioned in the link seems to be the result of multiplying Acc.z by thrustScale in the positionController and then adding the feedforward term. Is that correct?

https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller/position_controller_pid.c#L257

If so, it does not match the architecture shown in the diagram below.

![image](https://github.com/bitcraze/discussions/assets/143582301/2222d11e-2cc1-474a-b7c3-cbc4d07a8ea8)

I thought this was the thrust value obtained after calculating Attitude PID Control and Attitude Rate Control. Is it correct to say that the thrust value is the acceleration value converted to Thrust based solely on position and velocity?

**Reply by gemenerik** [reply]
> The control->thrust mentioned in the link seems to be the result of multiplying Acc.z by thrustScale in the positionController and then adding the feedforward term. Is that correct?

Yes, this is correct. However, the thrust you are referring to (a single float) does not directly equal the power distribution to the motors. This actually follows from (also) running the attitude and attitude rate PIDs.

**Reply by WoodyloveBo** [reply]
https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller/position_controller_pid.c#L260

Why is Thrustraw multiplied by Thrust Scale 1000? Thrustraw appears to be the acceleration value with PID gains applied. I want to know exactly what Thrust Scale represents.

**Reply by gemenerik** [reply, marked as answer]
From what I can tell it represents nothing more than a scaling of the value that comes out of the PID.
